src/Raven.Server/Documents/TimeSeries/TimeSeriesStorage.cs

```diff
@@ -82,10 +84,43 @@ public TimeSeriesStorage(DocumentDatabase documentDatabase, Transaction tx)
             tx.CreateTree(TimeSeriesKeysSlice);
         }
 
+        public void PurgeSegments(string collection, long upto, DocumentsOperationContext context)
+        {
+            var collectionName = _documentsStorage.ExtractCollectionName(context, collection);
+            var tableName = collectionName.GetTableName(CollectionTableType.TimeSeries);
+            var table = context.Transaction.InnerTransaction.OpenTable(TimeSeriesSchema, tableName);
+
+            if (table == null || table.NumberOfEntries == 0)
+                return;
+
+            var pendingDeletion = context.Transaction.InnerTransaction.CreateTree(PendingDeletionSegments);
+
+            using (context.Allocator.From(Bits.SwapBytes(upto), out var slice))
+            using (var it = pendingDeletion.MultiRead(collectionName.Name))
+            {
+                if (it.Seek(new Slice(slice)) == false)
+                {
+                    if (it.Seek(Slices.AfterAllKeys) == false)
+                        return;
+                }
+
+                do
+                {
+                    var etag = it.CurrentKey.CreateReader().ReadBigEndianInt64();
+                    if (etag > upto)
+                        continue;
+
+                    table.DeleteByIndex(TimeSeriesSchema.FixedSizeIndexes[CollectionTimeSeriesEtagsSlice], etag);
+                } while (it.MovePrev());
```

Comment:
What is we have 100000000 items here? Don't we need to break it up?

Reply:
This seems to be a more general problem even with regular tombstone and I guess we need the same behavior in both cases.
So I've opened an issue to address it: https://issues.hibernatingrhinos.com/issue/RavenDB-14484